### PR TITLE
[SPARK-41818][SPARK-42000][CONNECT] Fix saveAsTable to find the default source

### PR DESCRIPTION
**Pull request by ueshin**
### What changes were proposed in this pull request?

Fixes `DataFrameWriter.saveAsTable` to find the default source.

### Why are the changes needed?

Currently `DataFrameWriter.saveAsTable` fails when `format` is not specified because protobuf defines `source` as required and it will be an empty string instead of `null`, then `DataFrameWriter` tries to find the data source `""`.

The `source` field should be optional to let Spark decide the default source.

### Does this PR introduce _any_ user-facing change?

Users can call `saveAsTable` without `format`.

### How was this patch tested?

Enabled related tests.

### CODE REVIEW

File: python/pyspark/sql/readwriter.py [hunk 1]
```diff
@@ -468,7 +468,7 @@ def table(self, tableName: str) -> "DataFrame":
         |  8|
         |  9|
         +---+
-        >>> _ = spark.sql("DROP TABLE tblA")
+        >>> _ = spark.sql("DROP TABLE tblA").collect()
```
Comment (ueshin) [inline python/pyspark/sql/readwriter.py:471]: For [SPARK-41725](https://issues.apache.org/jira/browse/SPARK-41725).

File: python/pyspark/sql/connect/readwriter.py [hunk 2]
```diff
@@ -472,6 +472,8 @@ def save(
     def insertInto(self, tableName: str, overwrite: Optional[bool] = None) -> None:
         if overwrite is not None:
             self.mode("overwrite" if overwrite else "append")
+        elif self._write.mode is None or self._write.mode != "overwrite":
+            self.mode("append")
         self.saveAsTable(tableName)
```
Comment (ueshin) [inline python/pyspark/sql/connect/readwriter.py:477]: I guess we shouldn't reuse this for `insertInto`? Should revisit this.

**Review by HyukjinKwon — APPROVED** [review]
LGTM awesome!

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.4.

**Pull request closed without merging**